### PR TITLE
Decentralized well being game api

### PR DESCRIPTION
**Pull request by iumairazhar**
### Project Abstract

Well-being engagement is up and rising. With a more health-conscious generation of users, they are wanting to experience health engagement in different areas of their lives. Health Hero, as a digital health engagement company, seeks to grow a well-being-focused game API that engages developers with health-related development tools. With more people now working remotely, health engagement has never been more important. Healthcare is a multi-billion dollar industry that still struggles with delivering tools to consumers to engage with their health in their own way. IoT, omnichannel, chatbots, and other interactive interfaces are just a few to mention for the opportunities surrounding the use of Health Hero’s Game API

### For which grant [level](https://github.com/w3f/Grants-Program#level_slider-levels) are you applying? 
- [ ] **Level 1**:  Up to $10,000, 2 approvals
- [x] **Level 2**:  Up to $50,000, 3 approvals
- [ ] **Level 3**:  Unlimited, 5 approvals (for > $100k Web3 Foundation Council approval)

### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied, renamed ( `project_name.md`) and updated.
- [x] A BTC or Ethereum (DAI/USDT) address for the payment of the milestones is provided inside the application.
- [x] I have read and acknowledged the [terms and conditions](https://github.com/w3f/Grants-Program/blob/master/docs/T&Cs.md).
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process).

### CODE REVIEW

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the application. Just to double check: Should we close/terminate your previous application: https://github.com/w3f/Grants-Program/blob/master/applications/nft_product_analytics_suite.md We usually don’t allow teams to work on two grants at the same time. Additionally could you specify the ink! contract a little bit better (potential functionality). Otherwise it looks good to me.

**Comment by Noc2** [comment]
Done. Could you still specify the ink! contract a little bit better (potential functionality)?

**Comment by iumairazhar** [comment]
> Done. Could you still specify the ink! contract a little bit better (potential functionality)?

Updated thanks

**Review by Noc2 — DISMISSED** [review]
Looks good to me. Thanks for the update. I will share it with the rest of the team.

File: applications/decentralized_well-being_game_api.md [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+Decentralized Well-being Game API
+
+## Team Name
+
+Health Hero
+
+## Payment Address
+
+0xc7ad868f9b421dabf7cdaa0e6c508ff74e2f1cfd
```
Comment (semuelle) [inline applications/decentralized_well-being_game_api.md:13]: Could you add which currency you would like to be paid in (BTC, USDT, DAI or kUSD)?

**Review by semuelle — CHANGES_REQUESTED** [review]
1. Could you explain why you chose not to pursue your previous grant any further? 
2. What do you mean by "they will have the tools necessary to create customized smart contracts"? When do developers create their own smart contract and when do they interact with yours (mentioned in deliverable 1b)?
3. I'm confused about the architecture and the extent of this grant. Do I understand correctly that users submit data to your backend through some frontend (e.g. Teams)? And your backend triggers rewards through a smart contract? Where and how do other developers' smart contracts come in? An architecture diagram would be very helpful, and a list of smart contracts with exposed data and functions that you are planning to implement.

**Comment by iumairazhar** [comment]
1. Could you explain why you chose not to pursue your previous grant any further?
Health Hero came to a conclusion that the previous project scope/use case wouldn't be a solution that could be scalable for Health Hero and the market is already flooded with great NFT analytics solutions.
2. What do you mean by "they will have the tools necessary to create customized smart contracts"?
Developers would be able to customize the smart contracts to mold them around their specific use case with the ability to, for example, customize their own rewards program within the pre-defined smart contract provided by Health Hero.
3. When do developers create their own smart contract and when do they interact with yours (mentioned in deliverable 1b)?
Developers would create their smart contract which is inherited from the main Health Hero contract. At this point, they are able to then customize the contract based on their rewards program needs.
4. I'm confused about the architecture and the extent of this grant. Do I understand correctly that users submit data to your backend through some frontend (e.g. Teams)?
Yes, it would be from any chat and social media platform. Ex. Discord, Telegram, Microsoft Teams, Slack, WhatsApp, Messenger, Twitter, Instagram, TikTok, Etc.
5. And your backend triggers rewards through a smart contract?
Yes, it does.
6. Where and how do other developers' smart contracts come in?
The other developers' smart contracts come in when they want to customize the rewards layer based on their needs.

@semuelle I will be adding the proposed architecture diagram for better understanding.

Thanks

**Comment by semuelle** [comment]
> Developers would create their smart contract which is inherited from the main Health Hero contract. At this point, they are able to then customize the contract based on their rewards program needs.

That's helpful. Can you provide some info on what the structure of the base contract looks like?

> I will be adding the proposed architecture diagram for better understanding.

I will review again when you added the diagram.

**Comment by iumairazhar** [comment]
## Proposed  Architect Diagram
Screenshot 2021-11-12 at 11 42 53 PM
@semuelle  let me know if anything else you need

**Comment by semuelle** [comment]
Thanks. Could you add the diagram to the document, and also address the currency issue I mentioned above?

**Comment by iumairazhar** [comment]
> Thanks. Could you add the diagram to the document, and also address the currency issue I mentioned above?

Sure @semuelle

File: applications/decentralized_well-being_game_api.md (hunk 1 above)
Comment (iumairazhar) [reply applications/decentralized_well-being_game_api.md:13]: @semuelle  USDT thanks

Comment (semuelle) [reply applications/decentralized_well-being_game_api.md:13]: Into the document, please.

Comment (iumairazhar) [reply applications/decentralized_well-being_game_api.md]: updated thanks @semuelle

**Review by Noc2 — APPROVED** [review]

**Comment by semuelle** [comment]
@iumairazhar, I noticed that you are applying for USD 10,000, but checked Level 2 in your PR. Is that intentional?

**Comment by iumairazhar** [comment]
> @iumairazhar, I noticed that you are applying for USD 10,000, but checked Level 2 in your PR. Is that intentional?

sorry it was unintentional typo, we were meant to do for level 2 

@semuelle sorry for inconvenience

**Review by Noc2 — CHANGES_REQUESTED** [review]
Are you aware that you can also apply for a level 2 grant (including the announcement) and only ask for 10k. Based on the current level of details and delivery, 50k is in my opinion too much.

**Comment by semuelle** [comment]
I will have to agree with David here. Given the flow of data in your architecture, this API is not particularly decentralised. Thus, I find it of limited usefulness for the ecosystem. We are generally happy to support projects including ink! smart contracts, but more to the extent of 10,000 USD, not 50,000.

**Comment by iumairazhar** [comment]
@Noc2 @semuelle  so do I need to provide the diagram that show's it's decentralised right ?

**Comment by semuelle** [comment]
The architecture diagram you provided above contains a JS API, which, as I understand, receives data from clients and triggers transactions on the smart contract(s). That is not decentralised. If this is incorrect, please correct me.

**Comment by iumairazhar** [comment]
@semuelle that JS API is decentralized web3 which allow users to interact that was purposed architectures diagram. Let me know if that is causing confusing to you

**Comment by semuelle** [comment]
> Do I understand correctly that users submit data to your backend through some frontend (e.g. Teams)?
> >  Yes, it would be from any chat and social media platform. Ex. Discord, Telegram, Microsoft Teams, Slack, WhatsApp, Messenger, Twitter, Instagram, TikTok, Etc.
> 
> And your backend triggers rewards through a smart contract?
> >  Yes, it does.

You mentioned a backend receiving data from a client and triggering a smart contract. Now you are suggesting the clients execute transactions themselves. Perhaps you need to be more explicit in your description of the architecture and flow of information.

**Comment by iumairazhar** [comment]
> > Do I understand correctly that users submit data to your backend through some frontend (e.g. Teams)?
> > > Yes, it would be from any chat and social media platform. Ex. Discord, Telegram, Microsoft Teams, Slack, WhatsApp, Messenger, Twitter, Instagram, TikTok, Etc.
> > 
> > 
> > And your backend triggers rewards through a smart contract?
> > > Yes, it does.
> 
> You mentioned a backend receiving data from a client and triggering a smart contract. Now you are suggesting the clients execute transactions themselves. Perhaps you need to be more explicit in your description of the architecture and flow of information.


@semuelle @Noc2 Sorry For being not clear and causing confusion, Let me update the architecture diagram and flow of information in the application so it's more clear.

**Comment by semuelle** [comment]
Thanks for the update, @iumairazhar. 

If I'm reading the diagram correctly, the data goes straight from the client to the smart contract. How do you ensure that the data submitted is not faked? If my data triggers rewards, there is an incentive for users to submit fake data. Also, I just noticed that your deliverable specs mention state channels. Could you explain why state channels are needed and how you plan to implement them?

**Comment by iumairazhar** [comment]
> Thanks for the update, @iumairazhar.
> 
> If I'm reading the diagram correctly, the data goes straight from the client to the smart contract. How do you ensure that the data submitted is not faked? If my data triggers rewards, there is an incentive for users to submit fake data. Also, I just noticed that your deliverable specs mention state channels. Could you explain why state channels are needed and how you plan to implement them?


@semuelle 
We are going to implement KYC for the users and we are allowing only automated health activities using GoogleFit, Pleteon, Apple Health and there will be a proper threshold limit for each day.

We are planning to allow the users to do trading based on their earned coins using the state channels. For Example, if one user wants to sell some good like eatables, goodies, etc, another user can have those by offering his/her earning

**Review by Noc2 — CHANGES_REQUESTED** [review]
Do you already know how you are going to integrate state channels into ink! smart contract. This might be quite some additional work. You could potentially use https://github.com/perun-network/perun-polkadot-pallet and adapt it for ink! smart contract.

**Comment by iumairazhar** [comment]
> Do you already know how you are going to integrate state channels into ink! smart contract. This might be quite some additional work. You could potentially use https://github.com/perun-network/perun-polkadot-pallet and adapt it for ink! smart contract.

@Noc2 we are open to using this https://github.com/perun-network/perun-polkadot-pallet and adapting this our ink! smart contract.

**Review by semuelle — COMMENTED** [review]
Considering the relatively limited usefulness of the deliverables for the ecosystem and the lack of technical details, I agree with @Noc2 that 50,000 USD is too much. 10,000 would have been more appropriate in my view.

**Comment by Noc2** [comment]
> @Noc2 we are open to using this https://github.com/perun-network/perun-polkadot-pallet and adapting this our ink! smart contract.

Could you integrate this into the application or provide more information about your state channel solution?

**Comment by iumairazhar** [comment]
@Noc2  We will be allowing the users to earn tokens based on their authentic health data. Using these tokens they can buy stuff using state channels, We will also enable them to pay for their medicines/fruits/veggies, etc

I will be updating the application if need more details

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the reply. But as I said, it would be nice to have more technical information about the state channel solution as part of the grant application. Especially given the price point of the application (e.g. necessary ink smart contracts and the functions, backend for the state channel and the programming language, etc.).

**Comment by iumairazhar** [comment]
@Noc2 Working on this will share and update the application to show the state channels

thanks

**Comment by iumairazhar** [comment]
@Noc2 Backend for state channels will be in the rust language and we will be going to implement https://github.com/perun-network/perun-polkadot-pallet for state channels. These channels will allow us to make transactions to the users for their payouts from the treasury based on their authentic health data. As state channels will be established between treasury and users to handle frequently occurring transactions for their payouts.

State Channels (2)

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the update. So you will basically enable perun state channels for ink! smart contracts as part of the grant, correct? Could you integrate this into the application? Also it would still help a lot if you could reduce the price of the application. Otherwise, it doesn’t look like you will get the necessary number of approvals.

**Comment by iumairazhar** [comment]
> Thanks for the update. So you will basically enable perun state channels for ink! smart contracts as part of the grant, correct? Could you integrate this into the application? Also it would still help a lot if you could reduce the price of the application. Otherwise, it doesn’t look like you will get the necessary number of approvals.

@Noc2 sure doing that now

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks again for the update and reducing the price. However, in its current form it will be difficult to leverage the perun pallet from the ink! smart contract as far as I know. I believe you have to re-implement the perun pallet more or less as a smart contract in order to leverage it.

**Comment by iumairazhar** [comment]
@Noc2  We are open to re-implement the Perun pallet as part of an ink smart contract to leverage the state channels for the off-chain transactions from the treasury to the users.

**Comment by Noc2** [comment]
> @Noc2 We are open to re-implement the Perun pallet as part of an ink smart contract to leverage the state channels for the off-chain transactions from the treasury to the users.

Could you integrate this into the deliveries?

**Comment by iumairazhar** [comment]
@Noc2  Updated thanks

**Review by semuelle — APPROVED** [review]

**Review by Noc2 — APPROVED** [review]

**Review by BenWhiteJam — APPROVED** [review]

**Comment by iumairazhar** [comment]
@Noc2 Thanks for approving and supporting us so far Just to inform you we have started working on the project and keep you updated here and on emails as well.

Thanks and much love 
M Umair

**Pull request merged**